File: java/satellite/src/main/java/Satellite.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Satellite {

    private final List<Character> preorder;
    private final List<Character> postorder;
    private int current;

    public Satellite() {
        this.preorder = new ArrayList<>();
        this.postorder = new ArrayList<>();
        this.current = 0;
    }

    public Tree treeFromTraversals(List<Character> preorder, List<Character> postorder) {
        this.preorder.addAll(preorder);
        this.postorder.addAll(postorder);

        if (this.preorder.size() != this.postorder.size()) {
            throw new IllegalArgumentException("traversals must have the same length");
        }

        Set<Character> pre = new HashSet<>(preorder);
        Set<Character> post = new HashSet<>(postorder);

        if (!pre.equals(post)) {
            throw new IllegalArgumentException("traversals must have the same elements");
        }

        if (pre.size() != preorder.size() || post.size() != postorder.size()) {
            throw new IllegalArgumentException("traversals must contain unique items");
        }

        Node root = buildTree(0, postorder.size() - 1);

        return new Tree(root);
    }

    private Node buildTree(int left, int right) {

        if (left > right) {
            return null;
        }

        char root = this.preorder.get(this.current);
        this.current++;

        int middle = this.postorder.indexOf(root);

        Node node = new Node(root);

        node.left = buildTree(left, middle - 1);
        node.right = buildTree(middle + 1, right);

        return node;
    }
}

File: javascript/satellite/satellite.js
let current;
let preOrder;
let inOrder;

export const treeFromTraversals = (preorder, inorder) => {

  current = 0;
  preOrder = [...preorder];
  inOrder = [...inorder];

  if (preOrder.length !== inOrder.length) {
    throw new Error('traversals must have the same length');
  }

  let setPre = new Set(preOrder);
  let setIn = new Set(inOrder);

  let eqSet = (a, b) => a.size === b.size && [...a].every(value => b.has(value));

  if (!eqSet(setPre, setIn)) {
    throw new Error('traversals must have the same elements');
  }

  if (setPre.size !== preOrder.length || setIn.size !== inOrder.length) {
    throw new Error('traversals must contain unique items');
  }

  let root = buildTree(0, inOrder.length - 1);

  return root;
};

const buildTree = (left, right) => {

  if (left > right || current > preOrder.length) {
    return {};
  }

  let root = preOrder[current];
  current++;

  let middle = inOrder.indexOf(root);

  let node = {};
  node.value = root;

  node.left = buildTree(left, middle - 1);
  node.right = buildTree(middle + 1, right);

  return node;
};
